config/configsource/configsource.go

```diff
@@ -0,0 +1,267 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configsource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/spf13/viper"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+)
+
+const (
+	// ConfigSourcePrefix is used to identified a configuration source invocation.
+	ConfigSourcePrefix = "$"
+)
+
+// ApplyConfigSourcesParams holds the parameters for injecting data from the
+// given configuration sources into a configuration.
+type ApplyConfigSourcesParams struct {
+	// Set of configuration sources available to inject data into the configuration.
+	ConfigSources map[string]component.ConfigSource
+	// MaxRecursionDepth limits the maximum number of times that a configuration source
+	// can inject another into the config.
+	MaxRecursionDepth uint
```

Comment:
Is this needed now?

Reply:
Is there a danger of infinite recursion? If not then maybe this is not needed?

Reply:
It is already handled and tested in the initial code. We should support injecting the whole configuration and still having secrets to be resolved.